rust/src/composepost.rs

```diff
@@ -883,6 +883,35 @@ pub fn prepare_rpmdb_base_location(
     Ok(())
 }
 
+/// Recurse into this directory and return the total size of all regular files.
+pub fn directory_size(
+    dfd: i32,
+    mut cancellable: Pin<&mut crate::FFIGCancellable>,
+) -> CxxResult<u64> {
+    let cancellable = &cancellable.gobj_wrap();
+    let dfd = crate::ffiutil::ffi_view_openat_dir(dfd);
+    fn directory_size_recurse(d: &openat::Dir, cancellable: &gio::Cancellable) -> Result<u64> {
+        let mut r = 0;
+        for ent in d.list_dir(".")? {
+            cancellable.set_error_if_cancelled()?;
+            let ent = ent?;
+            let meta = d.metadata(ent.file_name())?;
```

Comment:
What about splitting the counting logic in this loop body right here, so that the filename can be recorded in error contexts for the remaining I/O accesses below?

Reply:
I don't understand quite what you mean by splitting - I added a `.with_context()` though!